RunningOrder/Shared/Utils/CloudKitContainer.swift

```diff
@@ -186,13 +105,120 @@ class CloudKitContainer {
+        guard !CloudKitContainer.createdCustomZone else { return }
+
+        Logger.verbose.log("shared zone creation")
+        let sharedZone = CKRecordZone(zoneID: ownedZoneId)
+        let zoneOperation = CKModifyRecordZonesOperation()
+        zoneOperation.recordZonesToSave = [sharedZone]
+
+        zoneOperation.modifyRecordZonesCompletionBlock = { _, _, error in
+            if let error = error {
+                Logger.error.log("error while creating custom zone : \(error)")
+            } else {
+                CloudKitContainer.createdCustomZone = true
+            }
+        }
+        cloudContainer.privateCloudDatabase.add(zoneOperation)
+    }
+
+    private init() {
+        createCustomZoneIfNeeded()
+        enableNotificationsIfNeeded(for: ownedZoneId)
+
+        if let owners = CloudKitContainer.ownerNames {
+            for (owner, _) in owners {
+                enableNotificationsIfNeeded(for: CKRecordZone.ID(zoneName: CloudKitContainer.zoneName, ownerName: owner))
+            }
+        }
+
+        Self.askPermissionForDiscoverabilityIfNeeded(in: cloudContainer)
             .store(in: &cancellables)
     }
 
-    func validateNotification(_ userInfo: [String: Any]) -> Bool {
-        guard let notification = CKNotification(fromRemoteNotificationDictionary: userInfo) else { return false }
+    func enableNotificationsIfNeeded(for zoneId: CKRecordZone.ID) {
+        let databaseToEnable = database(for: zoneId)
+        databaseToEnable.fetchAllSubscriptions()
+            .filter { $0.isEmpty }
+            .flatMap { _ in Self.createSubscriptions(for: databaseToEnable) }
+            .sink(receiveFailure: { error in Logger.error.log(error) })
+            .store(in: &cancellables)
+    }
+
+    func removeSubscriptions() {
+        [CKDatabase.Scope.private, .shared].forEach { scope in
+            let database = cloudContainer.database(with: scope)
+            database.delete(withSubscriptionID: database.subscriptionId) { (id, error) in
+                if let error = error {
+                    Logger.error.log("error at deletion of subscription : \(error)")
+                } else {
+                    Logger.debug.log("subscription deletion successful : \(String(describing: id))")
+                }
+            }
+        }
+    }
+
+    func test() {
+        cloudContainer.database(with: .private)
+            .fetchAllSubscriptions()
+            .print(in: .debug)
+            .sink(receiveFailure: { _ in }, receiveValue: { subs in
+                print(subs)
+            })
+            .store(in: &cancellables)
+
+        let fetchZones = CKFetchRecordZonesOperation(recordZoneIDs: [ownedZoneId])
+
+        fetchZones
+            .publisher()
+            .print(in: .debug)
+            .sink(receiveFailure: { _ in }, receiveValue: { _ in })
+            .store(in: &cancellables)
+
+        cloudContainer.database(with: .private).add(fetchZones)
+    }
+
+    // MARK: -
```

Comment:
Useless comment ?

Reply:
not useless, it allows to cut the file in the summary breadcrumb, and in the minimap